### ISSUE TITLE
journald (c|sh)ould support RFC5424 syslog on `/dev/log`

### ISSUE THREAD

**Issue by eqvinox** [systemd/systemd#19251]
**Is your feature request related to a problem? Please describe.**

No.

**Describe the solution you'd like**

When receiving messages over the `/dev/log` unix socket, it would be quite easy for `journald` to recognize RFC5424 syslog messages (they start with `<N>1 ` rather than `<N>Jan`.)  RFC5424 is much more modern (ISO8601 timestamps, structured key/value data support) than the modified (no hostname field, no linefeed) RFC3164 format previously used for `/dev/log`.

(For good measure, if RFC5424 is detected, it should probably be fully compatible, including line break handling - classic `/dev/log` does not expect linebreaks / multiple messages.)

**Describe alternatives you've considered**

The `sd_journal_send` family of functions is a systemd-specific alternative to this;  however for applications/daemons implementing key/value data it may be difficult/annoying to convert their k/v data to the varargs-based `sd_journal_send` API.  (This is because constructing `va_list`s in C is only possible through an actual function call - there is no way to build them programmatically.  If the k/v data already exists in another format, getting it into a `va_list` is at the very least extremely annoying if not impossible.)

**The systemd version you checked that didn't have the feature you are asking for**

Checked Debian's `247.3-3` version and looked at the 247 -> 248 git log & didn't see anything related.

**Comment by eqvinox** [comment]
(Addendum:)

The `sd_journal_send` family should probably also get a non-varargs based way to pass k/v data, but these 2 approaches are complementary, not exclusionary.  Since there is a standard for k/v data in syslog, why not support it?

(Yes, I'm aware the RFC is only intended for transport over TCP/UDP, and the `/dev/log` interface is in theory completely unspecified with the libc's `syslog()` being the "proper" entry point.  However, in practice, `/dev/log` is a UNIX socket with a well-understood format.)

**Comment by poettering** [comment]
RFC5424 is pretty weird, no? It's not actually used by anything but rsyslog, and only opt-in there? it seems like it's mostly a format private to rsyslog that just happens to have an RFC?

There's sd_journal_sendv() which takes an array of `struct iovec`, which should be pretty easy to use for putting together messgaes programmatically.

You can also re-implement the protocol sd-journal implements easily, though it currently has no nice docs. We could certainly open that up though. Which is requested in #17748

**Comment by eqvinox** [comment]
> RFC5424 is pretty weird, no? It's not actually used by anything but rsyslog, and only opt-in there? it seems like it's mostly a format private to rsyslog that just happens to have an RFC?

It's also supported by syslog-ng and reasonably widespread on network products (router, switches) as senders (which is tangentially relevant to systemd for receiving these messages on remote syslog?)

> There's sd_journal_sendv() which takes an array of `struct iovec`, which should be pretty easy to use for putting together messgaes programmatically.

Argh.  I misunderstood the API and thought the iovec was for composing the message text from multiple pieces.

> You can also re-implement the protocol sd-journal implements easily, though it currently has no nice docs. We could certainly open that up though. Which is requested in #17748

That has only minor benefits (not linking libsystemd?) over sd_journal_sendv.

It may still make sense to look at RFC5424 for `/dev/log` to get structured logs from anything that already implements RFC5424 and/or can't use sd_journal_sendv for whatever reason.

**Comment by poettering** [comment]
> It may still make sense to look at RFC5424 for `/dev/log` to get structured logs from anything that already implements RFC5424 and/or can't use sd_journal_sendv for whatever reason.

Dunno, our structure is quite different from RFC5424, and we have a pretty comprehensive default vocabulary of fields that don#t translate at all to what rsyslog does there. moreover, the protocol spoken via /dev/log is what glibc generates. I am not aware of anything trying to to send RFC5424 over that socket.

**Comment by poettering** [comment]
> That has only minor benefits (not linking libsystemd?) over sd_journal_sendv.

Well, libsystemd has no fallback on classic syslog if journal is not available, because noone cared about that. It might make sense for clients to prefer sd_journald_sendv() if it works, and fallback to syslog() if it doesnt.

**Comment by poettering** [comment]
BTW, I sat down and documented the natie protocol now. See https://systemd.io/JOURNAL_NATIVE_PROTOCOL

**Comment by eqvinox** [comment]
@poettering cool, Thanks!

I'll close this now, I don't think there's anything else to be done really.

**Comment by eqvinox** [comment]
@poettering  Oh, P.S.: `/dev/log` isn't RFC3164, it's RFC3164 without the hostname field (and without a trailing `\n`) - sending RFC3164 doesn't work :smile:

**Comment by eqvinox** [comment]
(Trailing `\n` if it's a `SOCK_STREAM`, no trailing `\n` if it's `SOCK_DGRAM`)

**Comment by poettering** [comment]
hmm, i figure we should beef up journald to accept RFC3164 compatible datagrams then. Unlike RFC5424 I think RFC3164 makes a ton of sense to support, since it describes what is actually used in the wild.

**Comment by eqvinox** [comment]
I don't think directly writing to `/dev/log` using full 3164 is actually common - I don't know of anyone doing it.  It doesn't work with other syslog daemons either, and it's not what the libcs do.

(On the other hand, both the extra hostname field and the trailing `\n` are easy enough to recongize and ignore...)

btw: random side question: since SOCK_DGRAM sockets are theoretically lossy, there might be some sense for journald to offer a SOCK_SEQPACKET in parallel?  The semantics are almost identical - it's "lossless" and guarantees in-order, with the downside of clients having to reconnect if journald restarts, and journald having to accept() and read from multiple fds...

**Comment by eqvinox** [comment]
Also, FWIW, the largest downside to 3164 style for any user aside from k/v support is the missing fractional timestamps — these days they're an extremely valuable debugging tool particularly for performance problems, and read-side timestamping is not a solution when the system is overloaded and introduces a delay :disappointed:

**Comment by eqvinox** [comment]
If the goal is to "improve the world", there are two wider-ranging things to do:

- add support for the journald socket to glibc and musl so it's used for `syslog()`, or alternatively create a LD_PRELOADable DSO that overrides `syslog()`
- create a new `/dev/log-5424` that accepts 5424 format messages and can also be implemented on the BSDs (and, again, should be supported in the libc) (of course another format could be used there as well)

**Comment by eqvinox** [comment]
(The reason for the `/dev/log-5424` is that clients otherwise don't know whether they can send 5424 format messages.  I didn't consider that side of the equation when I opened this PR; I had assumed a configuration option to do/not do 5424 on `/dev/log`, but that's not an option if `syslog()` is revamped.)

**Comment by eqvinox** [comment]
Oh and just to make this even more difficult… leap seconds! :laughing:

(These are a problem for 5424 too;  5424 mandates UTC.  The "correct" solution would be to use TAI, but that's hard enough to do on other POSIX systems and close to impossible on the zillions of closed-source syslog speakers.  Linux does have CLOCK_TAI, but a lot of systems don't configure the UTC-TAI offset correctly in the kernel, and also CLOCK_TAI is not accelerated in the vDSO and thus very slow.)

**Comment by poettering** [comment]
> btw: random side question: since SOCK_DGRAM sockets are theoretically lossy (and unordered), there might be some sense for journald to offer a SOCK_SEQPACKET in parallel?

SOCK_DGRAM on IP (i.e. UDP) is lossy. on AF_UNIX is reliable and fully ordered.

SOCK_DGRAM on AF_UNIX is awesome because it provides global ordering: it's a single queue and any program on the system writes to the same queue, thus you can sensibly relate log event by their order in it: what is queued earlier was submitted earlier, and what is queued later was submitted later.

SOCK_STREAM/SOCK_SEQPACKET implement independent queues for each client, and thus when you get woken up by epoll and start reading from one you might end up processing data that happened later than what is already queued elsewhere, but you'll only read later.

**Comment by poettering** [comment]
> Also, FWIW, the largest downside to 3164 style for any user aside from k/v support is the missing fractional timestamps — these days they're an extremely valuable debugging tool particularly for performance problems, and read-side timestamping is not a solution when the system is overloaded and introduces a delay disappointed

Well, either of the three syslog protocols uses calendar time, which is just borked...

**Comment by poettering** [comment]
> * add support for the journald socket to glibc and musl so it's used for `syslog()`, or alternatively create a LD_PRELOADable DSO that overrides `syslog()`

why though?

(musl upstream is not friendly towards systemd, and quite frankly I don't care about them. glibc is very conservative)

> * create a new `/dev/log-5424` that accepts 5424 format messages and can also be implemented on the BSDs (and, again, should be supported in the libc) (of course another format could be used there as well)

Why? Why would anyone want to submit RFC5424 unless rsyslog is behind /dev/log?

**Comment by poettering** [comment]
> (The reason for the `/dev/log-5424` is that clients otherwise don't know whether they can send 5424 format messages. I didn't consider that side of the equation when I opened this PR; I had assumed a client-side configuration option to do/not do 5424 on `/dev/log`, but that's not an option if `syslog()` is revamped.)

There are certainly other ways to communicate feature set for such a socket. For example an xattr on /dev/log or so that clients can check.

**Comment by eqvinox** [comment]
> SOCK_DGRAM on AF_UNIX is awesome because it provides global ordering: it's a single queue and any program on the system writes to the same queue, thus you can sensibly relate log event by their order in it: what is queued earlier was submitted earlier, and what is queued later was submitted later.

It's all great until you restart the receiver... then you suddenly get ENOTCONN or ECONNREFUSED without any prior notification.  In the middle of a multithreaded application down 30 levels of calls on the stack...

> Well, either of the three syslog protocols uses calendar time, which is just borked...
> 
> (the server side timestamps journald implicitly associates to all incoming messages via the _SOURCE_REALTIME_TIMESTAMP field is the kernel timestamp of the datagram submission btw, and typically preferrable over any client-side supplied calendar time, because accurate and trusted)

This is the why; while less trustworthy, client-side timestamps are quite badly limited by the current `/dev/log` protocol.  And the kernel timestamp doesn't help when the client batches log messages through `sendmmsg()`

> Why? Why would anyone want to submit RFC5424 unless rsyslog is behind /dev/log?

Well, for the common case, ISO 8601 timestamps with sub-second precision (or possibly TAI or monotonic, since this is a local interface), and for the rare case, k/v data.

> For example an xattr on /dev/log or so that clients can check.

I was assuming a `/dev/log` solution should work on the BSDs too.  I said "improve the world", not "improve Linux" :laughing:

**Comment by poettering** [comment]
> It's all great until you restart the receiver... then you suddenly get ENOTCONN or ECONNREFUSED without any prior notification. In the middle of a multithreaded application down 30 levels of calls on the stack...

Hmm? there's no difference between SOCK_DGRAM, SOCK_SEQPACKET, SOCK_STREAM on that. In fact, because we use SOCK_DGRAM things are a lot more robust if journald dies. How so? because the socket is actually bound by PID 1 and thus remains open even if journald is down, and gets passed back in on next invocation. This means you can actually restart journald any time, and can trivially easily continue with the next queued datagram and all this is entirely invisible to the client side. It's one of the joys of socket-based activation.

> Well, for the common case, ISO 8601 timestamps with sub-second precision (or possibly TAI or monotonic, since this is a local interface), and for the rare case, k/v data.

The former is only interesting if you actually intend to queue log datagrams locally in process for long periods of time before submitting them to syslog(). Why would you do that though? If you are afraid of blocking syslog(): enlarge the socket buffer. I mean, if you queue locally outside of the socket, what do you gain? memory isn't going to be less bounded than by the socket's own buffer?

and syslog() is not an API that allows k/v being passed in. Hence making syslog() speak RFC5424 or our native protocol isn't helping in anway on this, you can't get the structured data into syslog() in the first place.

> > For example an xattr on /dev/log or so that clients can check.
> 
> I was assuming a `/dev/log` solution should work on the BSDs too. I said "improve the world", not "improve Linux" laughing

I really don't care about BSD people, and I doubt they would even be interested a no BSD syslog implementation speaks rsyslog's RFC5424 to my knowledge.

**Comment by eqvinox** [comment]
> Hmm? there's no difference between SOCK_DGRAM, SOCK_SEQPACKET, SOCK_STREAM on that.

There is - you can get a notification of the disconnect and handle it in your main loop, and you know the receiver is up and alive when connect() succeeds.

> In fact, because we use SOCK_DGRAM things are a lot more robust if journald dies. How so? because the socket is actually bound by PID 1 and thus remains open even if journald is down, and gets passed back in on next invocation. This means you can actually restart journald any time, and can trivially easily continue with the next queued datagram and all this is entirely invisible to the client side. It's one of the joys of socket-based activation.

**Comment by eqvinox** [comment]
> I really don't care about BSD people, and I doubt they would even be interested a no BSD syslog implementation speaks rsyslog's RFC5424 to my knowledge.

Actually, FreeBSD has already been using RFC5424 for `/dev/log` since 3 years ago:

https://cgit.freebsd.org/src/commit/lib/libc/gen/syslog.c?id=8129693e78a67c2362bae5ad0d18e09e74eca311

**Comment by mbiebl** [comment]
@rgerhards any thoughts here? As author of RFC5424 I guess you have a better idea of how widely supported it is.

**Comment by eqvinox** [comment]
syslog-ng also accepts 5424 on `/dev/log` by default, since FreeBSD started doing it:

https://github.com/syslog-ng/syslog-ng/commit/17796b476d163774b03d5fc71370786badb8cf45

(the change affects all systems that use a datagram socket on `/dev/log`)

**Comment by eqvinox** [comment]
Reopening issue since after a bit more digging it seems actually useful.

**Comment by kaniini** [comment]
> (musl upstream is not friendly towards systemd, and quite frankly I don't care about them. glibc is very conservative)

I'm not sure how you've concluded this.  We would be happy to work with systemd, as long as the requests are reasonable.  Feel free to drop by #musl on freenode if you really want to talk about things.

**Comment by eqvinox** [comment]
Add.: NetBSD has used RFC5424 on `/dev/log` since late 2008:

https://github.com/NetBSD/src/commit/1c6aec2060baa7e97143b5df26d7e6bbeb409361

OpenBSD does its own thing.

**Comment by poettering** [comment]
> I'm not sure how you've concluded this.

https://ewontfix.com/14/
https://ewontfix.com/15/
https://twitter.com/richfelker/status/1184797548764114945
https://twitter.com/richfelker/status/644528616034988034
…

> We would be happy to work with systemd, as long as the requests are reasonable. Feel free to drop by #musl on freenode if you really want to talk about things.

Uh, I am not really on IRC.

Anyway, what's musl's take on RFC5424 then?

@codonell and what's yours? (or other glibc people)? Is there any desire to generate RFC5424 from libc syslog()? Apparently people are looking for that for the greater timer accuracy (though I don't really grok why that would be interesting, kernel timestamp sounds better to me in every way, and have been used by journal since it existed).

**Comment by codonell** [comment]
@poettering Yes, glibc is very conservative here, we have one RFC5424 (https://sourceware.org/bugzilla/show_bug.cgi?id=11715) related bug that is SUSPENDED (waiting for someone to present work done). I would be very hesitant to change the existing interfaces because of the expectations from consumers. We'd have to review *how* such a transition from one format to another would be conducted. We don't have anyone working on RFC5424 support. I have no opinion on RFC5424, other than it seems like someone documented something, but I don't know how useful that is given the current expectations from consumers. Someone, and I don't plan to do it, would need to look at existing consumers, their capabilities, and pipe the whole thing through and see what problems there are and that would be part of any analysis. It is a multi-component project. It's doable, but you have to get buy-in and justification early on in such a project.

**Comment by eqvinox** [comment]
Presumably the only real question here is how to detect RFC5424 support in the system.  First off, I think I should retract my statement about the detection method being desirable to work across platforms; since this is an internal-ish interface it's reasonable to have something Linux-specific here.

That said, xattrs are not a solution since you can't set xattrs on sockets.

I see 2 possible avenues of adding this:

- just implement a separate socket, e.g. `/dev/log-5424`.   Slightly ugly, and causes minor breakage in things like security rulesets which don't anticipate the new path.  Also breaks `SOCK_DGRAM` global ordering when both sockets are used.
  - worse sub-option: use an abstract path like `\0/dev/log` - this breaks when using network namespaces on the same mounts since abstract unix sockets are netns-local.
- make `/dev/log` a symlink and advertise 5424 support through a part of the path, e.g. `/dev/log` -> `/run/systemd/journal/dev-log-supports5424`.  Might also break some security rulesets, but does preserve global ordering.  (This would be my preferred option.)

In all cases, since RFC5424 and the old thing (pseudo-RFC3164) can coexist and even be mixed on a per-message basis, receivers need to (and already do, for `syslog-ng` - I haven't checked `rsyslog`) implement both.  This also means that the libc supporting it is a fully independent question, thought it would need to implement a fallback and couldn't just switch to RFC5424 exclusively.

**Comment by eqvinox** [comment]
@codonell since there's no way to get a "flag day" migration on Linux like FreeBSD & NetBSD did for this, it boils down to the syslog receiver somehow advertising support for 5424 and the libc's `syslog()` detecting and using it only if available.  Since the receiver is still forced to support pseudo-3164 in parallel, glibc can also just do nothing for a few years until the feature is *\*advertised\** more widely.

(*\*advertised\** because it's already *supported* by systems using syslog-ng, but there's no sensible way to detect that...)

**Comment by bazsi** [comment]
Original syslog-ng author here. After reading the thread above, I have a few notes to add, but TL;DR:

I am not sure RFC5424 format on /dev/log makes much sense. If it's about timestamps, there are easier solutions, especially if you are using syslog-ng :) (I am biased).

Now the details:

- rfc5424 is not an rsyslog "thing". For instance, I have nothing to do with rsyslog, but I was part of the specification process too. @rgerhards did contribute more to the text itself, but I would never say that it was created solely for the purposes of rsyslog. In syslog-ng, the RFC5424 driver is called "syslog()" which was my - failed - bet on RFC5424 being the future, in contrast, the RFC3164 format is using more technical names (like udp() and tcp() or lately network())
- rfc5424 has indeed started to see network adoption, I have personally seen samples of messages that used the format. Unfortunately many vendors get it wrong, but there are also those who get it right. RFC3164 is a horror though.
- I am not entirely sure that adopting 5424 for something like local logging makes that much sense, especially if it is not paired up with an API that allows name-value pairs to be specified.
- On the topic of RFC5424 and its improved timestamps: syslog-ng is able to restamp the incoming messages (not via SO_TIMESTAMP, rather simply by getting the current time at reception). It also ensures proper ordering. So just use keep-timestamp(no) for local sources. With that you can get micro-second resolution in the output syslog stream, simply using frac-digits(6)
- BSDs adopted RFC5424: not really sure about their motivation or what they wanted to solve and whether it turned out to be a good decision for them. 
- At one point in the past I did create eventlog, a library that had name-value pairs in its API and allowed the use of multiple log formats (structured like json/xml and unstructured like today's syslog). It even had an LD_PRELOADable DSO to override the normal "syslog()" interface. It never took of and I gave up a long time ago. systemd has such an interface for journald, but I have doubts it would be adopted that much.

With all that said, I think the role of RFC5424 is that the syslog implementation can augment the normal message with meta information that the application itself does not include in its log message, e.g. everything that journald has in underscore prefixed name-value pairs. And although syslog-ng could probably be configured to do that, it does not do so out-of-the box.

What I recommend application developers is that instead of using RFC5424 fields, they should adopt some easily parseable format (e.g. key=value or JSON with a preference on the first) and good naming conventions, and do all of this right within the text field provided by syslog. 

Reports, dashboards and all the niceties that build on top of syslog work on application specific use-cases anyway, so you don't have to come up with a "universal" naming scheme that would work across applications/domains. People tried coming up with taxonomies that would be adopted by all applications BUT IT HAS NEVER WORKED AND NEVER WILL. App developers can't get an ISO timestamp right, how would they pick the right fields to put information into?

**Comment by rgerhards** [comment]
sorry for the late reply - doing much SARS-CoV-2 volunteer work lately and that bites into my schedule.

I am the original author of rsyslog as well as have contributed quite a bit to RFC 5424.

All in all I tend to at least 90% with what @bazsi said.  Many of the things he said for syslog-ng, are also true:

* we support RFC5424 on the log socket, but you need to opt in
* we create high precision timestamps ourselves  (via SO_TIMESTAMP, an early suggestion by @poettering - thx!)
* the additional benefits for the log socket are questionable and mostly exists depending on proper structured data use IMHO

To the general theme:

* RFC5424 was a real IETF effort, many people have contributed to this effort, among others from Cisco and Huawei
* It is a network protocol, not more, not less
* RFC5424 is much more than just the format - you need to see it in combination with the transport RFCs. Unfortunately primarily RFC525 (*really* secure TCP/TLS) is in wider use, but not e.g. DTLS.
* A prime contribution of 5424 to the syslog field is this layered and extensible architecture
* Structured-Data was a compromise at a time there was no JSON but XML. XML was (is!) too verbose for syslog. If we would specify the format today, it would for sure be JSON.
* thus structured-data use has much been replaced by JSON. The benefits of better timestamp, security, more precise format etc. still exist and begin to get wider use. Remember how long it takes for IPv6 to get momentum ;-)

I am sure I forgot to mention some things. **All in all RFC5424 is best suited for real network scenarios with a lot of network gear and components involved.** For the log socket (as an IPC mechanism) it provides limited benefit.

Just my 2cts...

**Comment by eqvinox** [comment]
Thanks for your input Rainer & Balasz ... I think we can close this again here.

This does however leave me - specifically in the situation of an application developer who wants to submit k/v data to syslog - without a standard way to do that.  `journald` has its separate socket with its separate protocol, which is great because I can check if it exists, and if it does I just use it instead of `/dev/log`.

But what's the solution if I have `rsyslog` or `syslog-ng` on `/dev/log`?  How do I - *locally* - get messages with k/v data into my syslog?

**Comment by eqvinox** [comment]
(Continuing discussion about how to figure out moving to https://github.com/syslog-ng/syslog-ng/issues/3642 I guess, since it doesn't affect systemd/journald.)

**Comment by rgerhards** [comment]
> But what's the solution if I have `rsyslog` or `syslog-ng` on `/dev/log`? How do I - _locally_ - get messages with k/v data into my syslog?

Well, I would propose to use JSON inside the payload area, best with a "cookie" in front of it. This is how rsyslog supports it: https://www.rsyslog.com/files/temp/doc-indent/configuration/modules/mmjsonparse.html

I know that syslog-ng also supports it, but I don't have the details. But @bazsi and I worked jointly on the concept. I see this frequently in practice. My gut feeling is that this will finally win. (inside a RFC5424 container on the network)

**Comment by eqvinox** [comment]
@rgerhards that doesn't solve the detection problem... let me clarify; the problem isn't what format to use, it's how to do this without requiring the user to fiddle with the config first.

**Comment by bazsi** [comment]
You, as an application developer, should encode all your name-value pairs
into the syslog message, so it will work transparently over any syslog
implementation. You probably want to use the syslog severity levels, but
that's about it.

I would either use a key=value format (used widely in Unix logs, see
postfix as an example), or json. Just make absolutely sure, that it is easy
to identify the logs of your app without having to parse the structured
payload.

E.g. use a program name that identifies your software. You might also want
to add a prefix to the message field to indicate that it is json. (The
prefix ***@***.***: " is recognized by both rsyslog and syslog-ng).

People generally use these logs this way (e.g. in splunk):

* Search for the relevant logs (by program name or some other means)
* Once only the relevant logs are in focus, extract fields in an app
specific manner (usually via regexps or list of key=value, or json)
* Process this via whatever dashboarding, reports/analysis you need.

This means is that once your logs are easy to identify in the "syslog
soup", whatever format works that is easy to parse by
splunk/logstash/rsyslog/syslog-ng.

The only issue with stock syslogd, if the user uses that is the size limit
of the log message. Syslog-ng defaults to 64k, but this is configurable.
syslogd is hardwired to 1k messages max.

If your messages fit in 1k, you are done. If you need more, just require a
log implementation where this is configurable/higher by default, and which
uses tcp based transports (as udp would deteriorate quickly at
mtu-sizeof(udphdr) bytes).

If you want to provide your own log related functions within your
application (e.g not relying on a centralized log mgmt infrastructure),
then you might need to care about how the logging daemon needs to be
configured, otherwise just leave it to the user/log mgmt system.

**Comment by rgerhards** [comment]
full ACK to @bazsi

**Comment by eqvinox** [comment]
I can do that *if and when the user configures it*.

Meanwhile I'm trying to pass along the additional K/V data without user intervention where possible.  A JSON blurb in the message text - even appended after the plaintext message - would have some (not all) of my users show up at my door with torches and pitchforks.

My takeaway at this point is to just not do automatic K/V unless the journald socket is found/detected, in which case I can use the journald protocol.  I'm rather sad about that but I guess it is what it is.

Maybe rsyslog & syslog-ng will implement the journald socket at some point? :laughing:

**Comment by poettering** [comment]
> My takeaway at this point is to just not do automatic K/V unless the journald socket is found/detected, in which case I can use the journald protocol. I'm rather sad about that but I guess it is what it is.

Sounds reasonable to me. It's what Glib logging and others tend to do.

Sad, why so? I mean, you need a fallback path anyway. So if this means 1st try journal protocol, 2nd try classic syslog then that's not too bad?

> Maybe rsyslog & syslog-ng will implement the journald socket at some point? laughing

There's some rsyslog module somewhere that can read structured data from the journal as structured data.

The journal protocol's benefits to a large degree lie in the fact that its log records are well defined, and comprehensive (because we attach so many implicit metadata fields to it, and assign trust to it). Bypassing journald won't give you that.

There's also the issue that early-boot + initrd + late-boot stuff matters, and unless you intend to run rsyslog/syslog-ng there too you won't get it if you just listen on some socket. If you pull the data from the journal however, you have the benefit you won't lose that data, you can catch up from where you last read, and will be able to pick up whatever happened while you were offline.

Anyway, so my takeaways from this discussion:

1. On one hand: BSD supports RFC5424 on /dev/log since a while. Which I find interesting.
2. On the other hand: syslog-ng/rsyslog people see little value on doing RC5424 via /dev/log, since the only benefit appears to be more accurate timestamps, but there are better ways to have those. The RFC5424 is apparently more focussed on syslog-over-network.
3. There's no *nice* mechanism anyone came up with so far to allow feature negotiation on /dev/log, and without it RFC5424 on it is probably not feasible on Linux
4. My xattr idea for feature nego on /dev/log is not going to work. sniff. 

So, to me this appears, there's little actual reason to support RFC5424 in journald (or in glibc).

However, I guess full RFC3164 compat might make sense, simply because it's syslog baseline in a way. I added that to TODO.

And I guess we can close this here again

**Comment by rgerhards** [comment]
> There's some rsyslog module somewhere that can read structured data from the journal as structured data.

just for the record: it's `imjournal`, the default interface.

**Comment by poettering** [comment]
> That said, xattrs are not a solution since you can't set xattrs on sockets.

BTW, xattrs on socket inodes do work fine actually. Try this:

```sh
setfattr -n trusted.xxx -v yyy /run/systemd/journal/dev-log
```

what matters is that this is about the socket entry point inode, not the connected socket inode (i.e. the inode in the backing fs, not in sockfs). And there's also the slight problem that tmpfs only allows xattrs in the "trusted." xattr namespace, not in "user.". That means that it's not that great for feature nego, since it would only work for clients that have CAP_SYS_ADMIN.

**Comment by poettering** [comment]
> If there is no interest in systemd to support RFC5424, I might continue to use current format and just change to use gmtime() instead (without improved timestamps).

Note that consumers expect this to be in local time, and rfc3339 says so explicitly.

What I don't grok: local clients connecting to /dev/log will get SCM_TIMESTAMPNS anyway (and we consume that in journald). So why bother with the timestamps in the payload anyway? It would be useful with non-local senders, or with buffered senders, but /dev/log is generally not used like that, and glibc's client side definitely is not used like that. Hence what's the rationale for having better timestamps there?

**Comment by poettering** [comment]
btw, if anyone wants to do some feature nego on /dev/log, maybe one approach would be to encode some meta info in the /dev/log symlink. i.e. if you issue readlink(/dev/log) and it returns you a path, you could take basename() of it, and if that string contains a special marker (eg, a suffix of `.rfc5424`), then this could tell glibc that the receiver is capable of `.rfc5424`. if readlink() otoh fails because the thing is not a symlink, or if it contains another path, then this would mean only traditional baseline of bsd syslog is supported.

**Comment by eqvinox** [comment]
> What I don't grok: local clients connecting to /dev/log will get SCM_TIMESTAMPNS anyway (and we consume that in journald). So why bother with the timestamps in the payload anyway?

I can't speak to other use cases, but FRR buffers LOG_DEBUG level messages and batch submits them with `sendmmsg()`. (But then again, we do that because there's a lot of these, and in most cases people don't configure them to go to syslog if they enable it. But (then again)² since FRR is network control plane software, people do occasionally debug issues across multiple routers, where central syslog can be helpful. But (then again)³ if your routing is broken, your remote syslog may break too…)

My/our primary motivation was structured K/V data. We now do that on journald and manually-configured RFC5424.

> in systemd, we currently maintain a symlink /dev/log → /run/systemd/journal/dev-log. Should we one day actually implement .rfc5424 we would thus rename that socket (or hardlink it) to /run/systemd/journal/dev-log.rfc5424 and thus clients could identify with a single syscall which protocol they could speak.

ACK.

**Comment by zatrazz** [comment]
> > If there is no interest in systemd to support RFC5424, I might continue to use current format and just change to use gmtime() instead (without improved timestamps).
> 
> Note that consumers expect this to be in local time, and rfc3339 says so explicitly.

Users noted some issues about using localtime [1], but I don't have a strong preference here. glibc syslog format is already ad-hoc and it does not explicit adhere to RFC3339 now.

[1] https://sourceware.org/pipermail/libc-alpha/2021-March/123583.html

> 
> What I don't grok: local clients connecting to /dev/log will get SCM_TIMESTAMPNS anyway (and we consume that in journald). So why bother with the timestamps in the payload anyway? It would be useful with non-local senders, or with buffered senders, but /dev/log is generally not used like that, and glibc's client side definitely is not used like that. Hence what's the rationale for having better timestamps there?

Maybe because not everyone uses a server-side that implements SCM_TIMESTAMPNS or any other timestamp method. I see that making glibc side independent of server-side to provide better timestamp is a good improvement.

**Comment by poettering** [comment]
OK, so thinking about this more, here's what I'd propose if people really want RFC5425 support in glibc: change glibc to connect() to /dev/log.rfc5424 first. If that works, use it, and assume RFC5425 support. If it does not, fall back to /dev/log and stick to the classic syslog protocol glibc used so far. That should be sufficiently robust I think.

That said, I doubt we'd listen on that socket in journald, since afaics there's nothing valuable in rfc5425 for us... (you could interest me if glibc would send the TID and thread name along as implicit structured info, but otherwise there's nothing it would give us...)

**Comment by bazsi** [comment]
If that include process level information in addition to what we can get via /dev/log normally, I am interested (on behalf of syslog-ng).

Btw, syslog-ng accepts rfc5424 over /dev/log and has so for a while. IIRC I think it was added because of one of the BSDs.

**Comment by bazsi** [comment]
https://github.com/syslog-ng/syslog-ng/commit/17796b476d163774b03d5fc71370786badb8cf45

**Comment by fweimer-rh** [comment]
> OK, so thinking about this more, here's what I'd propose if people really want RFC5425 support in glibc: change glibc to connect() to /dev/log.rfc5424 first. If that works, use it, and assume RFC5425 support. If it does not, fall back to /dev/log and stick to the classic syslog protocol glibc used so far. That should be sufficiently robust I think.

I agree, I think using a different path is the best approach. But please do not use the `.rfc5424` suffix, but something else, in case the format isn't quite RFC-compliant. We wouldn't want to change our implementations in incompatible ways merely for RFC compliance.

**Comment by zatrazz** [comment]
> > Maybe because not everyone uses a server-side that implements SCM_TIMESTAMPNS or any other timestamp method. I see that making glibc side independent of server-side to provide better timestamp is a good improvement.
> 
> Well, maybe fix that syslog implementation then to use SCM_TIMESTAMPNS?

But the idea of improve glibc syslog timestamp is to make it independently whether server/relay support SCM_TIMESTAMPNS or the clock resolution used.

> 
> Sounds wrong to trying to "fix" glibc regarding this given it has a large chance of breaking glibc using programs all over the place, since they will generate logs that many syslog daemons can't parse? or will have a times in tiezones that servers don't expect?
> 
> From my perspective switching to RFC5424 or gmtime() in glibc without feature negotiation sounds like an exceptionally _really_ bad idea. You cannot just generate data on /dev/log that is not parsable or will be interpreted differently from the receivers than before. This is an API breakage I#d consider totally unacceptable.

Based on this discussion, it seems that RFC5424 does not have a widespread supports I would expect. So I have dropped it on my last patch iteration. Better timestamp is a nice thing to have, specially is will be with negligible overhead, but it is not the main drive of glibc syslog change. 

> 
> What I don't get though: apparently you are willing to add RFC5424 support to your syslog service, but adding SCM_TIMESTAMP support is off the table? I mean, either you can add new stuff to your syslog implementation or you cannot. If you can add one but not the other than sounds like a political mess you are in, and not sure why glibc (or systemd) should care about that?

I think you did not follow up the main driver of this change of glibc: essentially is to close the final bits of y2038 support and to improve timestamp precision. Former is relative simple, it just to use the internal y2038 interfaces, while for later I followed a standard instead of using another ad-hoc message formar. Personally I am more interested on fixing the y2038 support.

And I am proposing to use gmtime to handle the corner cases brought my on libc-alpha [2] and I does make sense to me to make glibc syslog to use UTC, even though it might characterize as breakage (other developers seems to be in line with this change as well). But I also don't have a strong preference here.

Finally, I am not against using SCM_TIMESTAMPNS, although it does incur in a lot of code boilerplate (to craft the sendmsg ancilliary data) using Linux-only tools (so it would only work on local relays). Embedding the timestamping in the message itself seems a way simpler solution.

[1] https://patchwork.sourceware.org/project/glibc/list/?series=8021
[2] https://sourceware.org/pipermail/libc-alpha/2021-March/123583.html

**Comment by bazsi** [comment]
On Mon, Mar 21, 2022, 20:42 Adhemerval Zanella ***@***.***>
wrote:

> Maybe because not everyone uses a server-side that implements
> SCM_TIMESTAMPNS or any other timestamp method. I see that making glibc side
> independent of server-side to provide better timestamp is a good
> improvement.
>
> Well, maybe fix that syslog implementation then to use SCM_TIMESTAMPNS?
>
> But the idea of improve glibc syslog timestamp is to make it independently
> whether server/relay support SCM_TIMESTAMPNS or the clock resolution used.
>
> Sounds wrong to trying to "fix" glibc regarding this given it has a large
> chance of breaking glibc using programs all over the place, since they will
> generate logs that many syslog daemons can't parse? or will have a times in
> tiezones that servers don't expect?
>
> From my perspective switching to RFC5424 or gmtime() in glibc without
> feature negotiation sounds like an exceptionally *really* bad idea. You
> cannot just generate data on /dev/log that is not parsable or will be
> interpreted differently from the receivers than before. This is an API
> breakage I#d consider totally unacceptable.
>
> Based on this discussion, it seems that RFC5424 does not have a widespread
> supports I would expect. So I have dropped it on my last patch iteration.
> Better timestamp is a nice thing to have, specially is will be with
> negligible overhead, but it is not the main drive of glibc syslog change.
>
> What I don't get though: apparently you are willing to add RFC5424 support
> to your syslog service, but adding SCM_TIMESTAMP support is off the table?
> I mean, either you can add new stuff to your syslog implementation or you
> cannot. If you can add one but not the other than sounds like a political
> mess you are in, and not sure why glibc (or systemd) should care about that?
>
> I think you did not follow up the main driver of this change of glibc:
> essentially is to close the final bits of y2038 support and to improve
> timestamp precision. Former is relative simple, it just to use the internal
> y2038 interfaces, while for later I followed a standard instead of using
> another ad-hoc message formar. Personally I am more interested on fixing
> the y2038 support.
>

I don't see how y2038 problem is affected by the timestamp generated by
syslog(), it does not contain a year in the first place.

Adding a year might be useful though and in that case using the rfc5424
timestamp format is a good choice.

Another alternative is to skip timestamp in its entirety: no need to
generate it, and no need to parse it, no y2038 problem and no timezone
changing issues.

I am not sure about other syslogd implentations but syslog-ng supports a
missing timestamp. Message would get timestamped as received or via
SCM_TIMESTAMP.

And I am proposing to use gmtime to handle the corner cases brought my on
> libc-alpha [2] and I does make sense to me to make glibc syslog to use UTC,
> even though it might characterize as breakage (other developers seems to be
> in line with this change as well). But I also don't have a strong
> preference here.
>
This would definitely break existing implementations, syslog-ng for sure. I
understand that gmtime() would resolve the timezone issue, but would
generate another, namely these messages would get interpreted with an
incorrect timestamp, especially if the timestamp format does not include
timezone offset, this would be bad.

And even if it does include a timezone offset, when aggregating logs, the
timezone may have informational value about the location of the system
producing the message.

I am strongly against non-local timezones in /dev/log

**Comment by zatrazz** [comment]
> I don't see how y2038 problem is affected by the timestamp generated by syslog(), it does not contain a year in the first place. Adding a year might be useful though and in that case using the rfc5424 timestamp format is a good choice. Another alternative is to skip timestamp in its entirety: no need to generate it, and no need to parse it, no y2038 problem and no timezone changing issues. I am not sure about other syslogd implentations but syslog-ng supports a missing timestamp. Message would get timestamped as received or via SCM_TIMESTAMP. And I am proposing to use gmtime to handle the corner cases brought my on

As I said, the timestamp is not the issue, but rather the internal interface glibc uses on syslog (which is easy to fix). The timestamping is another issue [1] that was set as SUSPENDED.

> libc-alpha [2] and I does make sense to me to make glibc syslog to use UTC, even though it might characterize as breakage (other developers seems to be in line with this change as well). But I also don't have a strong preference here.
> This would definitely break existing implementations, syslog-ng for sure. I understand that gmtime() would resolve the timezone issue, but would generate another, namely these messages would get interpreted with an incorrect timestamp, especially if the timestamp format does not include timezone offset, this would be bad. And even if it does include a timezone offset, when aggregating logs, the timezone may have informational value about the location of the system producing the message. I am strongly against non-local timezones in /dev/log

Yeah, I am inclined in dropping the gmtime change and keep on using localtime. Thanks for the input.

[1] https://sourceware.org/bugzilla/show_bug.cgi?id=11715

**Comment by poettering** [comment]
> But the idea of improve glibc syslog timestamp is to make it independently whether server/relay support SCM_TIMESTAMPNS or the clock resolution used.

I'd question your priorities here though. The number of applications using glibc, embedding glibc, carrying glibc in container images and so on vastly outnumber the number of syslog server implementations. Hence, focus on improving the few implementations, not the many: focus on fixing the syslog server implementations to use SCM_TIMESTAMP for example, instead of trying to fix a million apps/containers/… already out there. It's much easier to win fixing syslog servers then to win fixing millions of app installs...

**Comment by poettering** [comment]
> Finally, I am not against using SCM_TIMESTAMPNS, although it does incur in a lot of code boilerplate (to craft the sendmsg ancilliary data) using Linux-only tools (so it would only work on local relays). Embedding the timestamping in the message itself seems a way simpler solution.

Nope. You are mistaken. SCM_TIMESTAMP is implicit. The receiver enables it, not the sender. If a syslog implementation wants it, it can have it with zero client side changes. That's why it's so sexy: you get it basically for free, you do not have to change a single app to get it.

If syslog implementations can't get SCM_TIMESTAMP parsing from recvmsg() right, then I'd seriously question the quality of that syslog implementaiton. it's not precisely rocket science.

**Comment by poettering** [comment]
> I think you did not follow up the main driver of this change of glibc: essentially is to close the final bits of y2038 support and to improve timestamp precision. Former is relative simple, it just to use the internal y2038 interfaces, while for later I followed a standard instead of using another ad-hoc message formar. Personally I am more interested on fixing the y2038 support.

Hmm, the current glibc generated format does not carry any year information, unless I am mistaken? So it's kinda useless but that fact actually magically makes it y2038 compatible. Hence not following on the y2038 issue you are seeing?

**Comment by poettering** [comment]
> > 
> 
> I agree, I think using a different path is the best approach. But please do not use the `.rfc5424` suffix, but something else, in case the format isn't quite RFC-compliant. We wouldn't want to change our implementations in incompatible ways merely for RFC compliance.

If this is actually implemented, consider using some path in /run/ though, in some subdir. /dev is kinda for device nodes, backed by devtmpfs and putting something different there always sucked. That's why systemd actually makes /dev/log a symlink, because AF_UNIX sockets in devtmpfs is icky.

And putting it in a subdir is kinda nice, because that means the dir can be bind mounted easily and the socket in it still be replaced in a robust fashion without affecting the bind mount (i.e. bind mounting only the socket inode sucks, because if the owning process is restarted and deletes/rebinds the socket inode the bind mount is effectively gone). Hence, maybe put it in /run/syslog/socket or so.

**Comment by eqvinox** [comment]
@bazsi 

> If that include process level information in addition to what we can get via /dev/log normally, I am interested (on behalf of syslog-ng).

FRR includes the TID. Hadn't thought about thread name, need to check if we set those to useful values.

Apart from that you also get source file/line/function, a message identifier, a category code, and optionally format string arguments (i.e if the message was `syslog(prio, "foo %s bar %d", "abc", 123)` you get `[args@50145 arg0="abc" arg1="123"]`)

> Btw, syslog-ng accepts rfc5424 over /dev/log and has so for a while. IIRC I think it was added because of one of the BSDs.

It does indeed (I tested against syslog-ng), but there's no way to detect that on a Linux system sadly… so you have to manually put `destination syslog supports-rfc5424` in frr.conf :disappointed:

Would certainly appreciate some consensus method of autodetecting 5424 syslog, whether it be a different path, text in the symlink, xattr on the socket, or whatever else…

**Comment by zatrazz** [comment]
> > But the idea of improve glibc syslog timestamp is to make it independently whether server/relay support SCM_TIMESTAMPNS or the clock resolution used.
> 
> I'd question your priorities here though. The number of applications using glibc, embedding glibc, carrying glibc in container images and so on vastly outnumber the number of syslog server implementations. Hence, focus on improving the few implementations, not the many: focus on fixing the syslog server implementations to use SCM_TIMESTAMP for example, instead of trying to fix a million apps/containers/… already out there. It's much easier to win fixing syslog servers then to win fixing millions of app installs...

I noticed this some messages ago and I already have updated my patch on glibc to not move to RFC 5424.

> > Finally, I am not against using SCM_TIMESTAMPNS, although it does incur in a lot of code boilerplate (to craft the sendmsg ancilliary data) using Linux-only tools (so it would only work on local relays). Embedding the timestamping in the message itself seems a way simpler solution.
> 
> Nope. You are mistaken. SCM_TIMESTAMP is implicit. The receiver enables it, not the sender. If a syslog implementation wants it, it can have it with zero client side changes. That's why it's so sexy: you get it basically for free, you do not have to change a single app to get it.
> 
> If syslog implementations can't get SCM_TIMESTAMP parsing from recvmsg() right, then I'd seriously question the quality of that syslog implementaiton. it's not precisely rocket science.

I fully aware of this, I had the impression when you put that one should fix the syslog implementation you were considering glibc to implement SO_TIMESTAMPING to generate the timestamps at transmission time (instead when relay receives it).

> Hmm, the current glibc generated format does not carry any year information, unless I am mistaken? So it's kinda useless but that fact actually magically makes it y2038 compatible. Hence not following on the y2038 issue you are seeing?

It is a glibc implementation detail in fact, not really related to RFC5424.

**Comment by nuclight** [comment]
@bazsi,

> * At one point in the past I did create eventlog, a library that had name-value pairs in its API and allowed the use of multiple log formats (structured like json/xml and unstructured like today's syslog). It even had an LD_PRELOADable DSO to override the normal "syslog()" interface. It never took of and I gave up a long time ago. systemd has such an interface for journald, but I have doubts it would be adopted that much.

FreeBSD and Juniper made https://github.com/Juniper/libxo library used for both new and converting traditional utilities (`ls`, `wc`, `ps`, ...) to make XML/JSON output in addition to traditional human-readbale plain text. And `libxo` has `xo_syslog()` interface specifically supporting RFC 5424 format.

@zatrazz & others,

I don't understand why not to enable RFC 5424 in glibc and journald. My personal need these days was to test RFC 5424-compliant logging, which later could go onto network (precision timestamps & structured data, that's why), and test this locally, so - on `/dev/log`, yes. But journald does not support RFC 5424, and seems that glibc does not support it because of that. Chicken and egg problem. And either setting up virtual machine with BSD or installing alternative logging daemon in restricted environment - is a pain.

**Comment by eqvinox** [comment]
> For glibc we discussed this briefly and decided it is not worth because of the potential breakage on consumers that do not support RFC 5424. […] no straightforward way to advertise the syslog format […]

@zatrazz exactly this was a major point of discussion here — establishing some way to advertise that whatever is handling `/dev/log` is capable of processing RFC5424.

The option with the most agreement seemed to be requiring `/dev/log` to be a symlink with the feature indicator being in the symlink destination, e.g. `/dev/log -> /run/systemd/journal/dev-log-supports-5424`.

(Other options were an xattr on `/dev/log` or a completely new path like `/dev/log-5424`)

If no such agreed-upon indicator exists, I agree it's very painful to implement in glibc (or another syslog producer, e.g. FRR). We have a config option called `destination syslog supports-rfc5424` in FRR for now.

**Comment by poettering** [comment]
the xattr thing wouldn't work btw, becaus linux tmpfs does not allow user xattrs (only system/security xattrs so far) currently

**Comment by eqvinox** [comment]
> the xattr thing wouldn't work btw, becaus linux tmpfs does not allow user xattrs (only system/security xattrs so far) currently

Good to know. To be fair I only included the other options for the sake of completeness, I don't see anything speaking against the symlink based approach.

The only real TODO here is to agree on some rather basic things:

- what other people (if any) to involve in this consensus and how & where to document the result
- in fact using some text in the symlink target as indicator for RFC5424 support
- what depth are producers expected to hunt symlinks down for
- whether this is in fact RFC5424 messages as described in the RFC (remember "old-fashioned" `/dev/log` is *NOT* RFC3164 — it has the hostname field removed). Maybe the reference should be RFC5426? That's the UDP datagram encapsulation, one message per datagram, no extra cruft like message length or line feeds.
- the actual text to look for (`-rfc5424` if it is actual RFC5424 per previous item?)

I don't even care if anyone implements it — what I do care is "if you implement it, whoever knows when, do you agree with & will also do this feature indicator?"

If that agreement is there, I can put this in FRR, so there's going to be at least 1 producer for this. Which might help in breaking this chicken & egg situation.

**Comment by poettering** [comment]
upcoming kernel 6.6 adds support for xattr on tmpfs btw, so one could now attach a `user.rfc5426` xattr on /dev/log to indicate that the new rfc is supported, and then clients such as glibc could look for it, and if they find it send rfc5424.

**Comment by bazsi** [comment]
I am ok with adding the xattr to syslog-ng. I would rather use a reference to 5424 though. 5426 is a specific mapping (as are rfc5425 and rfc6587). `/dev/log` could be its own transport, if anyone would bother to write it up as an RFC.

I agree that in most ways it's similar to rfc5426 I just feel it's wrong to associate with a UDP specific mapping, as we are trying to use the same semantics (e.g. 5424) but with a custom transport (e.g. unix domain socket between the libc and the syslogd)

**Comment by poettering** [comment]
> I am ok with adding the xattr to syslog-ng. I would rather use a reference to 5424 though. 5426 is a specific mapping (as are rfc5425 and rfc6587).

sorry, i copied the wrong number from the discussions above. of course it should be named after the "main" RFC for all of this.

**Comment by bazsi** [comment]
I tried to implement this, but quoting fs/xattr.c

```
        /*
         * In the user.* namespace, only regular files and directories can have
         * extended attributes. For sticky directories, only the owner and
         * privileged users can write attributes.
         */
        if (!strncmp(name, XATTR_USER_PREFIX, XATTR_USER_PREFIX_LEN)) {
                if (!S_ISREG(inode->i_mode) && !S_ISDIR(inode->i_mode))
                        return (mask & MAY_WRITE) ? -EPERM : -ENODATA;
                if (S_ISDIR(inode->i_mode) && (inode->i_mode & S_ISVTX) &&
                    (mask & MAY_WRITE) && !inode_owner_or_capable(inode))
                        return -EPERM;
        }
```

And I get the EPERM error as I try to call setxattr() on the unix socket. I've verified that my code works on a regular file. So this mechanism will not work unfortunately.

As for the whys, this is what I have found in xattr(7):

```
       The file permission bits of regular files and directories are interpreted differently from the file permission bits of special files  and  symbolic
       links.  For regular files and directories the file permission bits define access to the file's contents, while for device special files they define
       access to the device described by the special file.  The file permissions of symbolic links are not used in access checks.  These differences would
       allow users to consume filesystem resources in a way not controllable by disk quotas for group or world writable special files and directories.

       For  this reason, user extended attributes are allowed only for regular files and directories, and access to user extended attributes is restricted
       to the owner and to users with appropriate capabilities for directories with the sticky bit set (see the chmod(1) manual page for an explanation of
       the sticky bit).
```

Back to the drawing board?

**Comment by poettering** [comment]
> Back to the drawing board?

Sniff. :-(

then the socket by a different name idea is probably the way to go.

**Comment by nuclight** [comment]
So any progress on this for two years?

**Comment by poettering** [comment]
@nuclight nope, noone submitted a change for the kernel to allow xattrs on afunix socket nodes. If this matters to you, consider submitting a patch yourself.

**Comment by poettering** [comment]
The kernel will likely soon have xattr support on AF_UNIX entry point sockets:

https://lore.kernel.org/all/20260216-work-xattr-socket-v1-0-c2efa4f74cb7@kernel.org/

**Comment by poettering** [comment]
btw, the kernel now supports attaching xattrs to sockets and socket entrypoint inodes. systemd has knobs for this already in .socket units, and the varlink socket transports we heavily use already attach xattrs to their sockets too.

hence, if there's still interest in what is proposed here we could now advertise the syslog support on the socket inodes themselves, as discussed. happy to review a patch.
